ROX-16381, ROX-16407: byodb connections

dashrews78 opened the pull request:
## Description

Setup to use an external database Postgres instance.  Requires the database be created.  

Summary of changes:
- if external do not restore a postgres backup bundle
- if external do not make any copies of the database.  ONLY use the database for the initial connection
- Switch the creation of the pg_stats extension to log as a warning instead of an error.
- fixed a rocks to postgres migration to not explicitly call out the `public` schema.  If a customer is using best practices they will have restricted use of the `public` schema.  (plus whenever we go to postgres 15 that will be restricted by default)
- As we phase out `central_previous` a lot of the things done here will become the main path.  Unfortunately we have to support the creation of `central_previous` as long as we support a rollback to 3.74 or 4.0.
- Do NOT attempt to calculate space for external databases

Changes to `central_previous` and `central_active` will occur under other tickets such as ROX-16774

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
~- [ ] Determined and documented upgrade steps~
~- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs]~(https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

Created a postgres sql database in GCP.  
Tested initial creation as well as upgrade from 3.73/3.74 RocksDB scenarios.

Ran a few iterations of access to ensure things worked.
- postgres user with postgres database with unrestricted permissions
- postgres user with stackrox database with unrestricted permissions
- stackrox user with stackrox database with unrestricted permissions
- stackrox user with stackrox database with restricted permissions

connection string used where user and database filled in from bullets above:
connectionString: "host=<MY_GCP_SQL_INSTANCE> user=<USER> statement_timeout=600000 client_encoding=SJIS database=<DATABASE>"

DB, user and permissions were created as follows:
```
# setup
drop database stackrox;
drop role readwrite;
drop user stackrox;

show search_path;

# create the user
  CREATE USER stackrox WITH PASSWORD 'THE_CREDS_I_WAS_USING';

# create database
create database stackrox;

# create schema for the user
\c stackrox
create schema stackrox;
# alter schema stackrox owner to stackrox;

# revoke rights on public
REVOKE CREATE ON SCHEMA public FROM PUBLIC;

# revoke rights on public to the database
REVOKE ALL ON DATABASE stackrox FROM PUBLIC;

# readwrite role
CREATE ROLE readwrite;
GRANT CONNECT ON DATABASE stackrox TO readwrite;
GRANT USAGE ON SCHEMA stackrox TO readwrite;
GRANT USAGE, CREATE ON SCHEMA stackrox TO readwrite;
GRANT SELECT, INSERT, UPDATE, DELETE ON ALL TABLES IN SCHEMA stackrox TO readwrite;
ALTER DEFAULT PRIVILEGES IN SCHEMA stackrox GRANT SELECT, INSERT, UPDATE, DELETE ON TABLES TO readwrite;
GRANT USAGE ON ALL SEQUENCES IN SCHEMA stackrox TO readwrite;
ALTER DEFAULT PRIVILEGES IN SCHEMA stackrox GRANT USAGE ON SEQUENCES TO readwrite;

# grant the readwrite role to stackrox
GRANT readwrite TO stackrox;
```

dashrews78 committed:
X-Smart-Branch-Parent: master

dashrews78 commented on CHANGELOG.md:
```diff
@@ -17,6 +17,8 @@ Please avoid adding duplicate information across this changelog and JIRA/doc inp
 ### Technical Changes
 
 - ROX-16962: A new parameter `spec.admissionControl.replicas` has been added to the `SecuredCluster` CRD.
+- Rollback to a 3.y release or the 4.0 release will no longer be supported starting from 4.3.
```
Took a stab with 4.3 open to suggestions here, but I think we need to make this statement soon.

connorgorman commented on migrator/clone/postgres/db_clone_manager_impl.go:
```diff
@@ -38,9 +39,55 @@ func New(forceVersion string, adminConfig *postgres.Config, sourceMap map[string
+
+	if d.versionExists(ver) {
+		// If the database version is newer than the software version make the user explicitly state they want to rollback.
+		// TODO:  Do we still want to do this `forceRollbackVersion` check?
```
I don't think we want the force rollback check unless we integrate it into the operator. Currently, we're working around this quite a bit in the operator code which does not handle this use case

dashrews78 commented on central/globaldb/postgres.go:
```diff
@@ -104,11 +104,13 @@ func InitializePostgres(ctx context.Context) postgres.DB {
-		activeDB := pgconfig.GetActiveDB()
+		if !pgconfig.IsExternalDatabase() {
+			// Get the active database name for the connection
+			activeDB := pgconfig.GetActiveDB()
```
Generally yes.  May take longer than we like as the `central_previous` etc have to remain until we no longer support a rollback to 4.0.

dashrews78 commented on central/main.go:
```diff
@@ -780,6 +780,12 @@ func notImplementedOnManagedServices(fn http.Handler) http.Handler {
 }
 
+func notImplementedWithExternalDatabase(fn http.Handler) http.Handler {
+	return utils.IfThenElse[http.Handler](
```
Oh I 100% agree.  This is what I get for copying other code.

connorgorman commented on central/main.go:
```diff
@@ -703,15 +703,15 @@ func customRoutes() (customRoutes []routes.CustomRoute) {
 			Route:      "/db/backup",
 			Authorizer: dbAuthz.DBReadAccessAuthorizer(),
-			ServerHandler: notImplementedOnManagedServices(
+			ServerHandler: notImplementedWithExternalDatabase(
```
perhaps not for this PR, but is there a way to get the secrets through the backups only? That may still be useful. something to think about

connorgorman commented on image/templates/helm/stackrox-central/values-public.yaml.example.htpl:
```diff
@@ -261,6 +261,10 @@
 #
+#     # External signifies that a Postgres wire-compatible database has already been deployed and a Central DB pod
+#     # does not need to be deployed
+#     external: false
```
Was this just missing from the docs?

dashrews78 replied:
yes, well from this example file

dashrews78 committed:
X-Smart-Squash: Squashed 16 commits:

d9b5706ba4 cut at external with single db
71c7255f71 additions
5383da7829 move external
2ea2f15f5b some adjustments to use external
2ec3fce269 debug
91f04f90f5 add flag to config map
465678fe39 dodge DB
410475c86d don't look at size
4c9140274e fix migration
c0add5cb99 use database name
ab37d15833 switch to a warning
5caf9effe1 remove schema
3ec9a425fc restore
ebb8709f66 cleanup
32136ad6e2 update change log
e10a59b692 style
